### DISCUSSION TITLE
how to set a function that is not use on params(not x, y, z and t)

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#23444] [Q&A General] [answered]
I meet a problem which i want to use variables(a, b, c) as a paramters of function. I find there only like this code
```C++
function_name->value(a);
// or
function.value(_t,  _q_point[_qp]);
```
when i want to set like this
```C++
function_name->value(a,b,c)
// or 
function_name .value(a,b,c)// this is wrong  obviously
```
make is fail
what should i do?

**Comment by GiudGiud** [comment]
You ~should~ could do something like:
function.value(_t,  Point(a, b, c);

However, when you define the function, you have to be mindful of what depends on t, and the Point, which is essentially represented by x, y, z

You ll need to be defining a quadri-dimensional function (or just use a ParsedFunction of x,y,z)

**Reply by YaqiWang** [reply]
Or you can create a user object that provides this function. If a, b, c are moose variables, you can couple them in the user object.

**Reply by hugary1995** [reply]
I agree with Yaqi. If your function has a fixed signature, just write an object for it. Depending on your specific need, it may be a user object, a material, a postprocessor, an aux kernel, etc. -- the flexibility here is the charm of MOOSE.

**Reply by GiudGiud** [reply, marked as answer]
Yeah even for a simple expression like 'a + b + c', using a material property like a ParsedMaterial is a better idea. 

If you are using AD, you must avoid using a function too unless you take care to call the ad routines on that function.
It is easier to just use material properties

**Reply by wangzhaohao** [reply]
Thanks all! i will try to use **userobject**.
